### ISSUE TITLE
Under closure: -O2,3 optimizations break when Module defined in pre-js

### ISSUE THREAD

**Issue by jcalfee** [emscripten-core/emscripten#5154]
The following breaks when attempting to use a Promise in `init.js`:
```bash
echo 'int main() {}' > main.cpp
EMCC_CLOSURE_ARGS="--language_in ECMASCRIPT6 --language_out ES5" emcc main.cpp -o function.html --pre-js "init.js" --closure 1 -O2
firefox function.html
```

>init.js
```javascript
var Module;
if (!Module) Module = (typeof Module !== 'undefined' ? Module : null) || {};

Module['initPromise'] = new Promise(resolve => {
//    Module['onRuntimeInitialized'] = function() {
//        console.log('Initialize');
//        resolve()
//    }
})
```

A Module object is available in the browser's console, however it appears to be the html's module version and does not have the emscripten methods (like Module.ccall).  The browser gets stuck saying "Downloading...".

If I run this without `-O2` it works fine.

**Comment by kripken** [comment]
Does it show any compilation errors? I think in -O2 we run the js optimizer, which might not be able to handle ES6 arrow functions, so I'd expect that to crash. `EMCC_DEBUG=1` in the env might show more.

**Comment by jcalfee** [comment]
No compilation errors.  Browser is suck "Downloading..." without the arrow function and Promise.

>init.js
```javascript
console.log('mod1', Module)
var Module;
if (!Module) Module = (typeof Module !== 'undefined' ? Module : null) || {};

//Module['initPromise'] = new Promise(resolve => {
//    Module['onRuntimeInitialized'] = function() {
//        console.log('Initialize');
//        resolve()
//    }
//})
```
Prints:
>mod1 undefined

Debug:
```
root@53d2f5cd844d:/src/test# EMCC_DEBUG=1 EMCC_CLOSURE_ARGS="--language_in ECMASCRIPT6 --language_out ES5" emcc main.cpp -o function.html --pre-js "init.js" --closure 1 -O2
DEBUG:root:PYTHON not defined in /root/.emscripten, using "/usr/bin/python"
DEBUG:root:JAVA not defined in /root/.emscripten, using "java"
DEBUG:root:Cache: acquiring multiprocess file lock to Emscripten cache
DEBUG:root:Cache: done
DEBUG:root:Cache: released multiprocess file lock to Emscripten cache
DEBUG:root:check tells us to use asm.js backend
WARNING:root:invocation: /usr/local/bin/emcc main.cpp -o function.html --pre-js init.js --closure 1 -O2  (in /src/test)
DEBUG:root:Checking JS engine ['nodejs']
INFO:root:(Emscripten: Running sanity checks)
DEBUG:root:compiling to bitcode
DEBUG:root:emcc step "parse arguments and setup" took 0.06 seconds
DEBUG:root:compiling source file: main.cpp
DEBUG:root:running: /clang/bin/clang++ -target asmjs-unknown-emscripten -D__EMSCRIPTEN_major__=1 -D__EMSCRIPTEN_minor__=37 -D__EMSCRIPTEN_tiny__=0 -D_LIBCPP_ABI_VERSION=2 -Werror=implicit-function-declaration -nostdinc -Xclang -nobuiltininc -Xclang -nostdsysteminc -Xclang -isystem/emscripten/system/include/libcxx -Xclang -isystem/emscripten/system/lib/libcxxabi/include -Xclang -isystem/emscripten/system/include/compat -Xclang -isystem/emscripten/system/include -Xclang -isystem/emscripten/system/include/SSE -Xclang -isystem/emscripten/system/include/libc -Xclang -isystem/emscripten/system/lib/libc/musl/arch/emscripten -Xclang -isystem/emscripten/system/local/include -O2 -std=c++03 -mllvm -disable-llvm-optzns main.cpp -Xclang -isystem/emscripten/system/include/SDL -emit-llvm -c -o /tmp/tmpJqyW9z/main_0.o
DEBUG:root:emcc step "bitcodeize inputs" took 0.01 seconds
DEBUG:root:optimizing main.cpp
DEBUG:root:emcc: LLVM opts: -O3 -disable-loop-vectorization -disable-slp-vectorization -vectorize-loops=false -vectorize-slp=false -vectorize-slp-aggressive=false  [num inputs: 1]
DEBUG:root:emcc step "process inputs" took 0.01 seconds
DEBUG:root:will generate JavaScript
DEBUG:root:including libc.bc
DEBUG:root:Cache: acquiring multiprocess file lock to Emscripten cache
DEBUG:root:Cache: done
DEBUG:root:Cache: released multiprocess file lock to Emscripten cache
DEBUG:root:including dlmalloc.bc
DEBUG:root:Cache: acquiring multiprocess file lock to Emscripten cache
DEBUG:root:Cache: done
DEBUG:root:Cache: released multiprocess file lock to Emscripten cache
DEBUG:root:emcc step "calculate system libraries" took 0.02 seconds
DEBUG:root:linking: ['/tmp/tmpJqyW9z/main_0_1.o', '/root/.emscripten_cache/asmjs/libc.bc', '/root/.emscripten_cache/asmjs/dlmalloc.bc']
DEBUG:root:emcc step "link" took 0.00 seconds
DEBUG:root:saving intermediate processing steps to /tmp/emscripten_temp
DEBUG:root:(not saving intermediate /tmp/emscripten_temp/emcc-0-basebc.bc because deferring linking)
DEBUG:root:emcc: LLVM opts: -strip-debug -disable-verify -internalize -internalize-public-api-list=main,malloc,free,__errno_location,fflush -globaldce -disable-loop-vectorization -disable-slp-vectorization -vectorize-loops=false -vectorize-slp=false -vectorize-slp-aggressive=false  [num inputs: 3]
DEBUG:root:emcc step "post-link" took 0.11 seconds
DEBUG:root:LLVM => JS
DEBUG:root:emscript: llvm backend: /clang/bin/llc /tmp/tmpJqyW9z/function.bc -march=js -filetype=asm -o /tmp/emscripten_temp/tmpHaFYjW.4.js -emscripten-global-base=8 -O2
DEBUG:root:  emscript: llvm backend took 0.0272858142853 seconds
DEBUG:root:emscript: js compiler glue
DEBUG:root:very limited syscalls (140, 6, 146, 54) so disabling full filesystem support
DEBUG:root:  emscript: glue took 0.241055011749 seconds
DEBUG:root:emscript: python processing: function tables and exports
DEBUG:root:asm text sizes[[102349, 2650], 579, 173, 1317, 0, 402, 385, 474, 268, 714, 725]
DEBUG:root:  emscript: python processing: function tables and exports took 0.00135207176208 seconds
DEBUG:root:emscript: python processing: finalize
DEBUG:root:  emscript: python processing: finalize took 0.000430822372437 seconds
DEBUG:root:emcc step "emscript (llvm => executable code)" took 0.28 seconds
DEBUG:root:applying pre/postjses
DEBUG:root:emcc step "source transforms" took 0.00 seconds
DEBUG:root:wrote memory initialization to function.html.mem
DEBUG:root:emcc step "memory initializer" took 0.00 seconds
DEBUG:root:running js post-opts
DEBUG:root:applying js optimization passes: asm eliminate simplifyExpressions simplifyIfs registerize minifyNames closure asmLastOpts last minifyWhitespace
chunkification: num funcs: 32 actual num chunks: 1 chunk size range: 106238 - 106238
DEBUG:root:config forcing native optimizer at /emscripten/optimizer
DEBUG:root:js optimizer using native
DEBUG:root:config forcing native optimizer at /emscripten/optimizer
.
running closure on shell code
DEBUG:root:closure compiler: java -Xmx1024m -jar /emscripten/third_party/closure-compiler/compiler.jar --compilation_level ADVANCED_OPTIMIZATIONS --language_in ECMASCRIPT5 --externs /emscripten/src/closure-externs.js --js /tmp/emscripten_temp/tmpmBlWmi.cl.js --js_output_file /tmp/emscripten_temp/tmpmBlWmi.cl.js.cc.js --externs /emscripten/third_party/closure-compiler/node-externs/http.js --externs /emscripten/third_party/closure-compiler/node-externs/vm.js --externs /emscripten/third_party/closure-compiler/node-externs/readline.js --externs /emscripten/third_party/closure-compiler/node-externs/domain.js --externs /emscripten/third_party/closure-compiler/node-externs/fs.js --externs /emscripten/third_party/closure-compiler/node-externs/tty.js --externs /emscripten/third_party/closure-compiler/node-externs/stream.js --externs /emscripten/third_party/closure-compiler/node-externs/events.js --externs /emscripten/third_party/closure-compiler/node-externs/child_process.js --externs /emscripten/third_party/closure-compiler/node-externs/https.js --externs /emscripten/third_party/closure-compiler/node-externs/url.js --externs /emscripten/third_party/closure-compiler/node-externs/string_decoder.js --externs /emscripten/third_party/closure-compiler/node-externs/dgram.js --externs /emscripten/third_party/closure-compiler/node-externs/repl.js --externs /emscripten/third_party/closure-compiler/node-externs/zlib.js --externs /emscripten/third_party/closure-compiler/node-externs/assert.js --externs /emscripten/third_party/closure-compiler/node-externs/dns.js --externs /emscripten/third_party/closure-compiler/node-externs/buffer.js --externs /emscripten/third_party/closure-compiler/node-externs/util.js --externs /emscripten/third_party/closure-compiler/node-externs/punycode.js --externs /emscripten/third_party/closure-compiler/node-externs/net.js --externs /emscripten/third_party/closure-compiler/node-externs/crypto.js --externs /emscripten/third_party/closure-compiler/node-externs/querystring.js --externs /emscripten/third_party/closure-compiler/node-externs/core.js --externs /emscripten/third_party/closure-compiler/node-externs/os.js --externs /emscripten/third_party/closure-compiler/node-externs/path.js --externs /emscripten/third_party/closure-compiler/node-externs/cluster.js --externs /emscripten/third_party/closure-compiler/node-externs/tls.js --externs /emscripten/third_party/closure-compiler/node-externs/process.js --language_in ECMASCRIPT6 --language_out ES5
DEBUG:root:emcc step "js opts" took 9.39 seconds
DEBUG:root:generating HTML
DEBUG:root:emcc step "final emitting" took 0.00 seconds
DEBUG:root:total time: 9.87 seconds
```

**Comment by kripken** [comment]
Is `EMCC_CLOSURE_ARGS` related here? Does the problem still exist without it?

**Comment by jcalfee** [comment]
Not related..  With the following:

`emcc main.cpp -o function.html --pre-js "init.js" --closure 1 -O2`

works:
```javascript
var Module; // eslint-disable-line no-var
//if (!Module) Module = (typeof Module !== 'undefined' ? Module : null) || {};
```
browser gets stuck "Downloading...":
```javascript
var Module; // eslint-disable-line no-var
if (!Module) Module = (typeof Module !== 'undefined' ? Module : null) || {};
```

**Comment by kripken** [comment]
Ok, looking into this, it seems like the problem is the HTML defines Module, and sets up the progress spinner and text area and so forth. Then that pre-js "kills" the Module object, replacing it, and then we don't have the hooks to update the progress spinner and its text area. So it gets stuck spinning forever with text "downloading..". main() has exectuted, though, and has printed to the web console (it doens't have the hook to print to the page's text area).

The issue is that if Module is defined in the HTML, then it's tricky in closure to check if it exists. I think you might need to use eval, see the top of src/shell.js. Better, though, would be to not check if Module exists, it always should. And you can assign to it on the HTML, by using --shell-file to use your own html.

**Comment by jcalfee** [comment]
Thank you, this helps me understand the environment..  

This module stuff is tricky business.  It would be nice to outsource this to projects like browserify and webpack.  I'm using webpack and will create separate node.js and web.js bundles  ..  Both web and node bundles can use `require(...)` and `module.exports = `.. 

Wren and the elliptic curve functions I'm compiling are not very small so closure would be nice to have.  Html and script tags are not important because anyone using these libs will use something like browserify or webpack to create or run the lib.  Webpack has about 5 options for output packaging types, so it is probably not something you want to re-implement.  Instead you could direct users to use these tools after you generate a module.export file especially if they want to use closure.

So, I'm really only after the raw object and a callback to know when it is ready to use.

If a simple output js could be:
`module.exports = function(done) {return {...`

Webpack or browserfy would handle any repackaging.

Another thing to consider, a --no-runtime flag could allow for a raw code bundle that would require your emscripten runtime. This would allow multiple libs (projects) to share the same runtime.
